packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -635,42 +725,79 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
     return null;
   }
 
-  function performUnitOfWork(workInProgress: Fiber): Fiber | null {
-    // The current, flushed, state of this fiber is the alternate.
-    // Ideally nothing should rely on this, but relying on it here
-    // means that we don't need an additional field on the work in
-    // progress.
-    const current = workInProgress.alternate;
-
-    // See if beginning this work spawns more work.
-    startWorkTimer(workInProgress);
-    if (__DEV__) {
-      ReactDebugCurrentFiber.setCurrentFiber(workInProgress);
-    }
-
-    let next = beginWork(current, workInProgress, nextRenderExpirationTime);
-    if (__DEV__) {
-      ReactDebugCurrentFiber.resetCurrentFiber();
-    }
-    if (__DEV__ && ReactFiberInstrumentation.debugTool) {
-      ReactFiberInstrumentation.debugTool.onBeginWork(workInProgress);
-    }
-
-    if (next === null) {
-      // If this doesn't spawn new work, complete the current work.
-      next = completeUnitOfWork(workInProgress);
-    }
-
-    ReactCurrentOwner.current = null;
+  let invokeBeginWork = beginWork;
+  if (__DEV__ && replayFailedBeginPhaseWithInvokeGuardedCallback) {
+    invokeBeginWork = function beginWorkInDEV(
+      current,
+      workInProgress,
+      expirationTime,
+    ) {
+      const stashedWorkInProgressProperties = Object.assign({}, workInProgress);
+      try {
+        return beginWork(current, workInProgress, expirationTime);
+      } catch (value) {
+        let thrownValue = value;
+        // Check if the thrown value is an error. If it is, we will replay the
+        // begin phase using invokeGuardedCallback, so that the error is not
+        // treated as caught by the debugger.
+        if (
+          value instanceof Error ||
```

Comment:
Let's inverse this check to look for thenables instead.

Reply:
I'll remove it entirely until the next PR